### PR TITLE
rex: add missing Perl resource

### PR DESCRIPTION
**Pull request by danielnachun**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Review by dawidd6 — DISMISSED** [review]

**Comment by danielnachun** [comment]
I can't reproduce this error locally on Big Sur:
```
cc -c  -I/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE -g -pipe -fno-strict-aliasing -fstack-protector-strong -DPERL_USE_SAFE_PUTENV -Os   -DVERSION=\"2.38\" -DXS_VERSION=\"2.38\"  "-I/System/Library/Perl/5.30/darwin-thread-multi-2level/CORE"   ReadKey.c
rm -f blib/arch/auto/Term/ReadKey/ReadKey.bundle
cc  -bundle -undefined dynamic_lookup -fstack-protector-strong  ReadKey.o  -o blib/arch/auto/Term/ReadKey/ReadKey.bundle  \
	      \	  

chmod 755 blib/arch/auto/Term/ReadKey/ReadKey.bundle
"/usr/bin/perl" "-Iblib/arch" "-Iblib/lib" ReadKey.pm.PL ReadKey.pm 
Can't load 'blib/arch/auto/Term/ReadKey/ReadKey.bundle' for module Term::ReadKey: dlopen(blib/arch/auto/Term/ReadKey/ReadKey.bundle, 0x0001): open("", O_RDONLY) failed with errno=2 at /System/Library/Perl/5.30/darwin-thread-multi-2level/DynaLoader.pm line 197.
 at /System/Library/Perl/5.30/darwin-thread-multi-2level/DynaLoader.pm line 95.
	DynaLoader::croak("Can't load 'blib/arch/auto/Term/ReadKey/ReadKey.bundle' for m"...) called at /System/Library/Perl/5.30/darwin-thread-multi-2level/DynaLoader.pm line 197
	DynaLoader::bootstrap("Term::ReadKey") called at ReadKey.pm.PL line 515
```
Maybe we should just use brewed `perl` instead?

**Comment by Bo98** [comment]
Isn't there a `Term::ReadKey` shipped with macOS that can be used? Same with a lot of these resources really - most can be Linux only unless there's hard version requirements.

**Comment by Bo98** [comment]
Please update to the latest version if you can. When doing so you can remove (for all OS):

```
List::MoreUtils
List::MoreUtils::XS
```

and replace with `List::Util` on macOS < Big Sur only (not even Linux).

I think `Data::Validate::IP` (and `NetAddr::IP` on Linux) is needed but that isn't new so maybe our test is missing coverage of that. `Digest::HMAC` (Linux only) is also another. `XML::Simple` has dependencies `XML::SAX` and `XML::SAX::Expat` not defined here. Try fix tests to cover all these first if possible as I might be wrong or have missed something.

Not sure if `inc::latest` is really needed.

---

These all ship with macOS and can probably be Linux only:

```
Clone::Choose (on Big Sur and later only)
Devel::Caller
Encode::Locale
Exporter::Tiny (on Big Sur and later only)
ExtUtils::MakeMaker
File::Listing
HTML::Parser
HTML::Tagset
HTTP::Cookies
HTTP::Daemon
HTTP::Date
HTTP::Message
HTTP::Negotiate
IO::HTML
IO::String
JSON::MaybeXS (on Big Sur and later only)
LWP::UserAgent
LWP::MediaTypes
Net::HTTP
PadWalker
Term::ReadKey
Text::Glob
Try::Tiny
URI
WWW::RobotRules
XML::NamespaceSupport
XML::Parser
XML::Simple
YAML (on Big Sur and later only)
inc::latest
```

**Comment by danielnachun** [comment]
@Bo98 I think I got all the changes in you requested so far.  There's a couple problems I'm still having on Linux and I want to see if they behave the same in CI:

1) `XML::Sax::Expat` fails to install claiming that it can't find its own module.  Maybe I've missed something obvious but it's currently got me a bit stumped.
2) If I comment that out, `XML::Simple` complains about the missing dependency but still installs, and `rex` itself tries to install.  It seems to "freeze" at the call to `perl Makefile.PL`, but I'm not sure if that wil happen in CI.

File: Formula/rex.rb [hunk 1]
```diff
@@ -12,186 +12,224 @@ class Rex < Formula
...
+    resource "JSON::MaybeXS" do
+      url "https://cpan.metacpan.org/authors/id/E/ET/ETHER/JSON-MaybeXS-1.004003.tar.gz"
+      sha256 "5bee3b17ff9dcffd6e99ab8cf7f35747650bfce1dc622e3ad10b85a194462fbf"
+    end
 
-  resource "Encode::Locale" do
-    url "https://cpan.metacpan.org/authors/id/G/GA/GAAS/Encode-Locale-1.05.tar.gz"
-    sha256 "176fa02771f542a4efb1dbc2a4c928e8f4391bf4078473bd6040d8f11adb0ec1"
-  end
+    resource "Scalar::List::Utils" do
```
Comment (danielnachun) [inline Formula/rex.rb:38]: The way this is currently implemented, all the resources in this block get installed on Linux.  That's mostly desired, except for `Scalar::List::Utils`, which you mentioned is only need on macOS before Big Sur, but not Linux.  Is it okay to wrap put this inside an `on_macos` block inside of the `if MacOS.version` block?  I'm happy to organize this differently if there a better way.

**Comment by danielnachun** [comment]
This is the error that has me stumped for `XML::Sax::Expat`:
```
==> perl Makefile.PL INSTALL_BASE=/home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec
Checking if your kit is complete...
Looks good
Generating a Unix-style Makefile
Writing Makefile for XML::SAX::Expat
Writing MYMETA.yml and MYMETA.json
==> make PERL5LIB=/home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib:/home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5
cp Expat.pm blib/lib/XML/SAX/Expat.pm
Manifying 1 pod document
==> make install
Can't locate XML/SAX/Expat.pm in @INC (you may need to install the XML::SAX::Expat module) (@INC contains: /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5 /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5 /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0 /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/Cellar/perl/5.34.0/lib/perl5/5.34.0 /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.34.0/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.34.0) at /home/linuxbrew/.linuxbrew/Cellar/rex/1.13.4/libexec/lib/perl5/XML/SAX.pm line 147.
Makefile:792: recipe for target 'install_sax_expat' failed
```

**Comment by danielnachun** [comment]
@Bo98 can you take a look at this?

**Comment by Bo98** [comment]
Oops missed the ping here somehow. I'll have a look later tonight/tomorrow.

**Pull request closed without merging**